README.md
# data-quality-checker

The data training quality checker for AIOps log training

## Prequisite

* Install dependencies
  * `pip install requirements/base.txt`
  * `pip install requirements/test.txt` for running unit test
* Set ENV
  * Add the path of this repo in `PYTHONPATH`, e.g. `export PYTHONPATH=/Users/ME/data-quality-checker:$PYTHONPATH`

## Run the data checker
  * Usage:

  ```
    Usage: checker.py [OPTIONS] FILE

    Options:
      -n, --sample-size INTEGER  SAMPLE_SIZE: Number of samples. Default:1000
      -r, --random-sampling      Randomly check SAMPLE_SIZE line of logs
      -f, --full                 Run checks for all the logs, the results are out of order.
      -p, --parallel INTEGER    Number of parallelism for running the full check. Default:4
      --help                     Show this message and exit.
  ```

  * Default: `python checker.py TEST_LOGS_FILE >results`
    * The fist `SAMPLE_SIZE` lines will be checked and reported if there is any problem found.
    * The `SAMPLE_SIZE`  can be set by `-n`. e.g. `-n 10000`
    * It runs faster when pipe the output to a file.
  * Random sampling: `python -r checker.py TEST_LOGS_FILE >results`
    * `SAMPLE_SIZE` lines will be sampled from the input file if it has more than `SAMPLE_SIZE` lines. These lines will be checked and reorted if there is any problem found.
  * Full scanning: `python -f checker.py TEST_LOGS_FILE >results`
    * Check the all the logs in the file. Results will be not be in the original order.
    * The number of parallism can be set by `-p`. e.g. `-p 8`
    * For reference, checking 10M logs takes around 5 minutes with default parallism on a 2019 Mac Book Pro.

## Run unit test
  * `pytest logs_data_checker_tests -s --cov=logs_data_checker --cov-report term-missing --cov-report cov.xml`




## Contributing
Pull requests are very welcome! Make sure that your patches are tested. Ideally create a topic branch for every separate change you make. For example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

**Note:** Make sure you update the [Changelog](CHANGELOG.md) each time you add, remove, or update sample files.

## License & Authors

If you would like to see the detailed LICENSE click [here](LICENSE).

- Author: New OpenSource IBMer <new-opensource-ibmer@ibm.com>

```text
Copyright:: 2019- IBM, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.0.1] - 2019-02-15
- Project began

## [0.0.2] - 2020-07-24

### Added
- Added cp4mcm-cleanup-utility.sh; it aids the customer in removing CP4MCM resources
- Added ma-prereq.sh; it aids the customer in installing Mutation Advisor prereas
- Added ma-prereq-offline.sh; it aids the customer in installing Mutation Advisor prereas for offline case.
- Added cp4mcm-ha-utility.sh; it aids the customer to enable HA for CP4MCM

[unreleased]: https://github.com/ibm/repo-template/compare/v0.0.1...HEAD
[0.0.1]: https://github.com/ibm/repo-template/releases/tag/v0.0.1

## [0.0.3] - 2020-07-28

### Added
- Added the following three yaml files that are used for Unified Agent deployment:
  - `reloader.yaml`
  - `ua-deploy.yaml`
  - `ua-operator-deploy.yaml`

## [0.0.4] - 2020-09-30

### Added
- Added legacy search monitoring job deployment script
  - `scripts/monitor-mcm`

## [0.0.5] - 2020-10-20

### Added
- Added cloud native monitoring uninstall script
  - `scripts/cnmon/uninstall_cnmon.sh`

## [0.0.6] - 2020-10-22

### Added
- Added RHACM 2.1 issuer workaround script
  - `scripts/cp4mcm-rhacm21-cp-issuer-secret.sh`
  
## [0.0.6] - 2020-10-24
- Added uninstall.sh; it is a standalone utility to help the customer uninstall the Cloud Pak for MultiCloud Management; valid for use in CP4MCM 2.0
  - `scripts/uninstall.sh`
  
## [0.0.7] - 2020-10-27
- Removed default arguments to uninstall.sh
  - `scripts/uninstall.sh`
  
## [0.0.8] - 2020-11-30
- Updated cp4mcm-cleanup-utility.sh script to be more generic
  - `scripts/cp4mcm-cleanup-utility.sh`

## [0.0.9] - 2020-12-17
- Added grc-policy-updates.sh script to update GRC policy attributes when upgrading from mcm core 2.0 to Red Hat Advanced Cluster Management version 2.1
  - `scripts/grc-policy-updates.sh`

## [0.10.0] - 2021-01-29
- Added chatops-update.sh script to update Chatops integration settings
  - `scripts/chatops-update.sh`

## [0.11.0] - 2021-02-11
- Added CouchDB backup/restore scripts
  - `scripts/backupcouch.sh`
  - `scripts/restorecouch.sh`

## [0.11.1] - 2021-04-30
- Updated grc-policy-updates.sh script to be consistent with RHACM policy changes.
  - `scripts/grc-policy-updates.sh`

## [0.11.2] - 2021-05-03
- Added vaultfeeder tool to configure HashiCorp Vault and push target VM credentials in it for IBM CloudPak for Multicloud Management VM policy controller 2.0 onwards.
  - `vaultfeeder`

## [0.12.0] - 2021-05-25
- Updated the uninstall and cleanup scripts for ChatOps removal
  - `scripts/cp4mcm-cleanup-utility.sh`
  - `scripts/uninstall.sh`

## [0.13.0] - 2021-08-26
- Updated cp4mcm-ha-utility.sh script to support modules
  - infrastructure-management-grc
  - infrastructure-management-vm

## [0.14.0] - 2022-05-26
- Updated cp4mcm-cleanup-utility.sh and uninstall.sh scripts to include commands to delete MutatingWebhookConfiguration and ValidatingWebhookConfiguration objects
  - `scripts/cp4mcm-cleanup-utility.sh`
  - `scripts/uninstall.sh`
  
## [0.15.0] - 2022-12-07
- Updated cp4mcm-ha-utility.sh by removing everything that dealt with removed operands for CP4MCM 2.3 FP6
- Added function to delete sre clusterrole and clusterrolebinding
  - `scripts/cp4mcm-ha-utility.sh`
  - `scripts/cp4mcm-cleanup-utility.sh`


logs_data_checker/checker.py
#
# IBM Confidential
# OCO Source Materials
#
# Licensed Materials - Property of IBM
#
# 5737-M96
# (C) Copyright IBM Corp. 2020 All Rights Reserved.
#
# The source code for this program is not published or otherwise
# divested of its trade secrets, irrespective of what has been
# deposited with the U.S. Copyright Office.
#
# US Government Users Restricted Rights - Use, duplication or
# disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
#

import argparse
import calendar
import datetime
import json
import logging
import math
import os
import random
import time
from concurrent import futures
from itertools import islice
from pathlib import Path

import click
import dateutil
import fastjsonschema  # for faster validation per https://www.peterbe.com/plog/jsonschema-validate-10x-faster-in-python
from dateutil.parser import parse as date_parse
from importlib_resources import files

from logs_data_checker import resources

LOG_SCHEMA_FILE = "log_schema.json"
logger = logging.getLogger(__name__)
BLOCK_SIZE = 10000


class FileInBatches(object):
    """ Iterator to load the file by batches
        each batch has BLOCK_SIZE lines

    """

    def __init__(self, file_path: Path):
        self.__file = open(file_path, "r", encoding="utf-8")
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.__file.closed:
            raise StopIteration
        line = list(islice(self.__file, BLOCK_SIZE))
        if line:
            self.count += 1
            return line, (self.count - 1)
        else:
            self.__file.close()
            raise StopIteration


def n_line(file):
    """ Return the number of lines for a file

    Args:
        file (str): File path

    Returns:
        int: number of lines
    """
    count = 0
    buffer = 16 * 1024 * 1024  # buffer=16MB
    with open(file, "rb") as inf:
        while True:
            block = inf.read(buffer)
            if not block:
                break
            count += block.count(b"\n")
    return count


class VerificationResult(object):
    """ Result of a check
    """

    def __init__(self, error_type: str, message: str):
        self.error_type = error_type
        self.message = message


def validate_normalized(line):
    """ Validate the log in json with respects to the schema

    Args:
        line (str): log in json format

    Returns:
        VerificationResult: if any errors, else None
    """
    try:
        validator(line)
    except fastjsonschema.JsonSchemaException as e:
        return VerificationResult("[Validation Error]", e.message)
    return None


def check_timestamp_digits(timestamp):
    """ Check the timestamp format. We expect 13 digits (to milliseconds) timestamp

    Args:
        timestamp (int): timestamp in epoch

    Returns:
        VerificationResult: if any errors, else None
    """
    if isinstance(timestamp, int) and int(math.log10(timestamp)) + 1 == 13:
        return None
    return VerificationResult("[Timestamp Error]", "Incorrect timestamp")


def unix_time_millis(time_str):
    """ Parse the time str and covert to epoch in millisecond
    e.g.
        unix_time_millis("2020-01-10 15:12:15.680000") = 1578669135000

    Note that the parser support only the seconds. The epoch is naively converted to millisecond.

    Args:
        time_str (str): Time string

    Returns:
        int: epoch in millisecond
    """
    dt = date_parse(time_str)
    return calendar.timegm(dt.timetuple()) * 1000


def time_sync_tolerance(time1: int, time2: int, tor: int = 1000):
    """ Comapre timestamp and utc_timestamp to seconds

    Args:
        time1 (int): [description]
        time2 (int): [description]
        tor (int, optional): [description]. Defaults to 5.

    Returns:
        bool: True if diff is under tor
    """
    if abs(time1 - time2) < tor:
        return True
    return False


def check_time(line):
    """ Check the consistency between line["timestamp"] and line["utc_timestamp"]

    Args:
        line (Dict): a line of log of json format, parsed as a dict

    Returns:
        VerificationResult: if any errors, else None
    """
    if "timestamp" not in line:
        return VerificationResult("[Time Error]", 'Missing "timestamp"')
    if "utc_timestamp" in line:
        try:
            assert time_sync_tolerance(
                line["timestamp"], unix_time_millis(line["utc_timestamp"])
            )
        except dateutil.parser._parser.ParserError as e:  # pylint: disable=protected-access
            return VerificationResult("[Time Error]", str(e))
        except AssertionError as e:
            return VerificationResult(
                "[Time Error]", '"timestamp" and "utc_timestamp" are not consistent'
            )
    return None


def line_check(line):
    """ Run checking for a line

    Args:
        line ([type]): [description]

    Returns:
        [type]: [description]
    """
    json_line = json.loads(line)

    errors = [
        "%s: %s" % (e.error_type, e.message)
        for e in [
            validate_normalized(json_line),
            check_timestamp_digits(json_line["timestamp"])
            if "timestamp" in json_line
            else None,
            check_time(json_line),
        ]
        if e
    ]

    return errors


with open(files(resources) / LOG_SCHEMA_FILE) as inf:
    schema = json.load(inf)
validator = fastjsonschema.compile(schema)


def _pcheck(block):
    """ Helper for multiprocesses: check a block of logs

    Args:
        block List[List[str], int]: lines, block_id

    Returns:
        [type]: [description]
    """

    results = []
    lines, block_id = block

    for li, line in enumerate(lines):
        json_line = json.loads(line)
        result = [
            "%s: %s" % (e.error_type, e.message)
            for e in [
                validate_normalized(json_line),
                check_timestamp_digits(json_line["timestamp"])
                if "timestamp" in json_line
                else None,
                check_time(json_line),
            ]
            if e
        ]
        global_line_number = block_id * BLOCK_SIZE + li
        results.append((global_line_number, result))
    return results


def parallel_checking(file: Path, parallism: int):
    """ Check the file in parallel

    Args:
        file (Path): Input file
        parallism (int): number of processes
    """
    batches = FileInBatches(file)
    with futures.ProcessPoolExecutor(max_workers=parallism) as executor:
        tasks = {
            executor.submit(_pcheck, block): (block_index, block)
            for block_index, block in enumerate(batches)
        }
    for future in futures.as_completed(tasks):
        errs = future.result()
        for line_cnt, e in errs:
            if len(e) > 0:
                print("Line %d has %d errors: %s" % (line_cnt, len(e), ", ".join(e)))


@click.command()
@click.argument("file", nargs=1, type=click.Path(exists=True))
@click.option(
    "-n",
    "--sample-size",
    default=1000,
    type=int,
    help="SAMPLE_SIZE: Number of samples. Default:1000",
)
@click.option(
    "-r",
    "--random-sampling",
    default=False,
    is_flag=True,
    help="Randomly check SAMPLE_SIZE line of logs",
)
@click.option(
    "-f",
    "--full",
    default=False,
    is_flag=True,
    help="Run checks for all the logs, the results are out of order",
)
@click.option(
    "-p",
    "--parallel",
    default=4,
    type=int,
    help="Number of parallelism for running the full check",
)
def main(file, sample_size, random_sampling, full, parallel):
    """ Entry point for the checker

    Args:
        file (str): input file
        sample_size (int): Number of samples. Default:1000
        random_sampling (bool): Randomly check $sample_size line of logs. Default:False
        full (bool): Run checks for all the logs, the results are out of order. Default:False
        parallel ([type]): Number of parallelism for running the full check
    """
    if full:
        parallel_checking(file, parallel)
        return

    sampled_logs = []
    file_cnt = n_line(file)
    if sample_size > file_cnt:
        random_sampling = False

    if random_sampling:
        samples = set(random.sample(range(0, file_cnt), sample_size))
        cnt = 0
        with open(file, "r", encoding="utf-8") as inf:
            for i in inf:
                if cnt in samples:
                    sampled_logs.append(i)
                cnt += 1
        samples = sorted(list(samples))
    else:
        cnt = 0
        with open(file, "r", encoding="utf-8") as inf:
            for i in inf:
                sampled_logs.append(i)
                cnt += 1
                if cnt == sample_size:
                    break
        samples = range(sample_size)

    for line_num, line in zip(samples, sampled_logs):
        errors = line_check(line)
        if len(errors) > 0:
            print(
                "Line %d has %d errors: %s" % (line_num, len(errors), ", ".join(errors))
            )


if __name__ == "__main__":
    main()
